### PR TITLE
feat: add rest transport generation for clients with optional flag

### LINKED ISSUE #267: Re-add HTTP transport and make it the fallback
Some users may not want to use gRPC. Http is a sufficient fallback for certain APIs, so this transport must be explicitly supported.
This support used to exist but was removed due to testing limitations.
See https://github.com/googleapis/gapic-generator-python/pull/114 for additional details.

### PR DESCRIPTION
**Pull request by yon-mg**
This pr adds traditional rest/json transport to GAPIC clients. As of now, this does not yet handle all cases and there may be bugs in the generated client. As such, there is now a --transport flag set to 'grpc' by default so any new additions won't be reflected in generated clients.

### CODE REVIEW

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:46]: This should be taken or computed from the service.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:103]: I know that this is a pattern in the existing templates, but IMO reaching into `self.__dict__` and liberal use of `setattr` and `getattr` is only justified when doing something complicated or performance sensitive. I think I'd prefer it if there was instead 'operations_client_' attribute that is initialized to `None` in the ctor if the service has LRO, and then lazily instantiated/returned in this property.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:114]: This is possibly a matter of opinion, but I think it would be better to have this logic live in the `Method` class. It makes it easier to unit test.
Also, what if there's more than one 'http' option? Is it deliberate that we pick the last one?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:109]: Is namespace a jinja class?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:108]: +1 for using enums.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:157]: I'm leaning towards anything that involves state modification living in the generator datatypes. It's easier to test and easier to change if we need to.

**Review by software-dov — COMMENTED** [review]
First pass, some broad outlines and some nitpicking.

I'm more in favor of naming the new transport and its module "Http" instead of Rest; it is possible to do protobuf over http, and it's more immediately obvious to someone swimming in the code what it means.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:46]: Oops I accidentally left this in there. Will revert.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:114]: By 'live in the method class' do you mean adding another field to the class?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:109]: Yes. It allows variable assignment to carry through a loop body since apparently just using the `{%- set %}` tag and a variable loses any changes once it leaves for loop scope.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:157]: I'm not sure I understand what you mean by generator datatypes. Could you clarify?

**Comment by busunkim96** [comment]
Is the REST transport going to be added in one PR? If not, can we maintain a long-running branch and merge to it instead? 

The microgenerator is now relied on by Cloud clients at GA stability so I'd prefer not to to surface this until it is somewhat stable.

**Comment by yon-mg** [comment]
> Is the REST transport going to be added in one PR? If not, can we maintain a long-running branch and merge to it instead?
> 
> The microgenerator is now relied on by Cloud clients at GA stability so I'd prefer not to to surface this until it is somewhat stable.

This sounds reasonable to me. I'll defer to the other reviewers on what the best practice should be. @software-dov @vam-google @vchudnov-g

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:114]: Oh, my mistake, I didn't read line 110 carefully. Yes, there can be only one option under a specific name.
Yes, I think adding another field or accessor method would be better.
Is 'http' an options extension? If so, the more idiomatic way of accessing it is shown with resources.
Strawman example:
```python
def http_opt(self) -> Optional[str]:
    opt = self.options.Extensions[http.http_opt]
    return opt[1] if opt else None
```

If this isn't an options extension, then ignore the above example.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:109]: Neat! TIL

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (software-dov) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:157]: The `gapic/schema/wrappers.py:{MessageType,Service,Method}` classes. The auxiliary datastructures that the generate creates before iterating over the templates.

**Comment by software-dov** [comment]
If it's not too much work, I think I'd be more in favor of feature toggles and small PRs instead of having a long-running branch. It's easier to do merging. If that seems like too much work, I'm okay with a long-running branch.

**Comment by yon-mg** [comment]
> I'm more in favor of naming the new transport and its module "Http" instead of Rest; it is possible to do protobuf over http, and it's more immediately obvious to someone swimming in the code what it means.

I agree with this somewhat but for consistency across other languages, flag arguments and regapic design specs, I think it's more appropriate to name it Rest. In addition, it's because it's possible to do protobuf over http that naming it such may be ambiguous. Rest seems to me more clearly defined.

File: gapic/utils/options.py
Comment (software-dov) [inline gapic/utils/options.py:43]: +1 for using an enum

File: gapic/utils/options.py
Comment (software-dov) [inline gapic/utils/options.py:141]: Can you add a quick comment indicating the expected format of the 'transport' option and maybe an example?

File: gapic/generator/generator.py
Comment (software-dov) [inline gapic/generator/generator.py:278]: This little logic chunk could be lifted up and generalized, maybe something like
```python
def is_undesired_transport(template_name: str, transport: str, opts) - > bool:
    return transport in template_name and transport not in opts.transport
```

Also: I'm still leaning towards "http" instead of "rest" for the template naming scheme. It just seems more descriptive of what the protocol is.

**Review by software-dov — COMMENTED** [review]
Looks good, just a few things.

File: gapic/generator/generator.py
Comment (software-dov) [inline gapic/generator/generator.py:305]: This is such a common pattern that python has a builtin function for determining it.
```python
return any(transport in template_name for transport in desired_transports)
```

File: gapic/schema/wrappers.py
Comment (software-dov) [inline gapic/schema/wrappers.py:745]: Strictly speaking this breaks the return signature. It would be more consistent to either change the return to be `Optional[Dict[str, str]]` or change the default value of answer to be an empty dictionary.

File: gapic/generator/generator.py
Comment (software-dov) [inline gapic/generator/generator.py:279]: Can you put parentheses around this chunk, just to make it clear that it's a cohesive unit?

File: gapic/utils/options.py
Comment (software-dov) [inline gapic/utils/options.py:17]: Is `Set` used somewhere?

**Review by software-dov — COMMENTED** [review]
LGTM, just two minor nits.

**Review by software-dov — APPROVED** [review]

File: gapic/ads-templates/%namespace/%name/%version/__init__.py.j2
Comment (vam-google) [inline gapic/ads-templates/%namespace/%name/%version/__init__.py.j2:23]: Please use your github id for TODOs (here and in the other places)

File: gapic/schema/wrappers.py
Comment (vam-google) [inline gapic/schema/wrappers.py:811]: http GET requests are automatically considered idempotent because they do not change data on the server. Post requests are usually non-idempotent, because it is usually unsafe retry write operation if it failed, because it may lead for example into writing the same information twice on the server.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2
Comment (vam-google) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:226]: won't `if not self._operation_client` work? I thought it was more idiomatic for python.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2
Comment (vam-google) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:178]: Why was this changed (looks unrelated to your changes)?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc_asyncio.py.j2
Comment (vam-google) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc_asyncio.py.j2:231]: Why are you changing this code (looks unrelated to http transport stuff)?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2
Comment (vam-google) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/rest.py.j2:175]: Please add new line

**Review by vam-google — APPROVED** [review]
LGTM
All the comments are minor. Please address them and then push the changes (please use squash and merge button). Thanks @software-dov for sharing your expertise in Python review!

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [inline gapic/schema/wrappers.py:755]: For clarity I would suggest 
```
body_spec = http[1]
answer[body_spec[0].name] = body_spec[1]
```

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [inline gapic/schema/wrappers.py:750]: Based on our offline discussion, let's s/`'method'`/`'verb'`/, since "method" is overloaded (service method, object method, HTTP method)

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [inline gapic/schema/wrappers.py:748]: Based on our offline discussion, I suggest s/`http_method`/`primary_binding`/ for clarity.

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [inline gapic/schema/wrappers.py:751]: Consider having "template" in the name: maybe `'path_template'` would be more descriptive and accurate.

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [inline gapic/schema/wrappers.py:744]: I do think a type declaration or a comment saying this is a `List[Tuple[FieldDescriptor, str]]` would be helpful.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2
Comment (vchudnov-g) [inline gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:34]: This seems like a non-trivial stylistic change, from having the transport name in a variable to constructing it in the template. I'm OK if @software-dov is fine with it, but just calling it out to make sure this is intentional.

File: gapic/schema/wrappers.py
Comment (yon-mg) [reply gapic/schema/wrappers.py:811]: I get the GET is idempotent but couldn't there be methods without an http annotation that are also idempotent? are grpc only methods inherently not so?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:178]: Typo

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc_asyncio.py.j2
Comment (yon-mg) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc_asyncio.py.j2:231]: https://github.com/googleapis/gapic-generator-python/pull/688#discussion_r516194162
I decided to update the other templates as well for consistency.

File: gapic/utils/options.py
Comment (vchudnov-g) [inline gapic/utils/options.py:54]: Maybe clarify this is a `+`-separated list of transports....

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2
Comment (vchudnov-g) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:178]: I approve of fixing typos on the fly!

File: gapic/schema/wrappers.py
Comment (vchudnov-g) [inline gapic/schema/wrappers.py:743]: In this comment, consider showing the expected output structure (like you did in our chat, which was very helpful)

**Review by vchudnov-g — APPROVED** [review]
Looks good. Only minor comments. Nice job! Very excited to have this in!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2
Comment (software-dov) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:226]: It's a little bit complicated. Some values can be falsy even if they are non-null, e.g. empty lists and strings.
In this particular case, it's slightly more idiomatic to do `if self._operations_client`, but I personally am fine with explicit checks in cases like this.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc_asyncio.py.j2
Comment (software-dov) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc_asyncio.py.j2:231]: +1 for consistency

File: gapic/schema/wrappers.py
Comment (software-dov) [reply gapic/schema/wrappers.py:750]: If this be madness, there be method to it! 💀

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2
Comment (software-dov) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:34]: I'm more in favor of keeping it the way it was, with 'GrpcTransport' provided by the service's `grpc_transport_name` property.

**Pull request merged**
